src/QMCWaveFunctions/RotatedSPOs.cpp

```diff
@@ -18,14 +18,17 @@
 
 namespace qmcplusplus
 {
-RotatedSPOs::RotatedSPOs(std::unique_ptr<SPOSet>&& spos)
+RotatedSPOs::RotatedSPOs(std::unique_ptr<SPOSet>&& spos, bool creatingCopy)
     : SPOSet(spos->isOMPoffload(), spos->hasIonDerivs(), true),
       Phi(std::move(spos)),
       nel_major_(0),
       params_supplied(false)
 {
   className      = "RotatedSPOs";
   OrbitalSetSize = Phi->getOrbitalSetSize();
+  // Ensure that a copy of the SPO data is only made once
+  if (!creatingCopy)
```

Comment:
There needs to be detailed documentation, its quite cryptic.  I guess that sometimes the caller of this constructor might have another copy of the object with the unique_ptr and that other times they don't so they going to invoke the maybe noop storeParams call.  It's difficult to reason about, and it seems like it should have to be at this layer at all.

What I really think is:

Several things are going wrong here.  One is that RotatedSpoSet wants to both compose and inherit from SPOSet.  This is a bad smell and probably means not just this class is unclear but so is the base class. May there is a data structure and a call interface here they probably don't both need to be type erased. Until that is cleared up the rest is going to be pretty confusing.  But I'm going to assume that needs to be a different PR.

It also seems like some versions of the derived SPOSet that is composed can storeParams and others can't / don't.  This makes reasoning about this code unhappy. We are managing a state machine that only exists in one of the cases.  Calls to do this have polluted the base class API, flags for this appear in the composing class.  Something needs to be seriously refactored or redesigned here.

Reply:
I don't think inheriting from SPOSet and storing an SPOset is the problem.  Conceptually, orbital rotation is modifying an existing set of SPO's, so modeling it that way in the code reflects the problem domain.  The variable Phi stores the original SPOSet.
What might help is creating another class that derives from SPOSet and introduces applyRotations and related code.  Maybe call it ModifiableSPOSet (or RotatableSPOSet).  And then the SPOSet's that support rotation (LCAO and Splines) can derive from that.

Reply:
What might be more clear is just have a function that applies the rotation and returns a new SPOSet that has been rotated.

Isn't this to preserve a weird implementation detail where the optimizer always calculates the total rotation from the original SPOSet?  Getting what is essentially a new set of Coefficients returned means you can make sure that surprising behavior is know where it might needed to reason about why the rotation at each application is what it is. 
```c++
 SPOSet RotateOrbitals(const SPOSet& sposet)
```
then later I read code like this
```c++
sposet_opt = OptimizingRotator.rotate(rotation, sposet_orig)
```
```c++
//or if your update of the rotation as I would expect
sposet_opt =  OptimizingRotator.rotate(delta_rotation, sposet_opt)
```

My understanding is that SPOSet is too heavy for this to be feasible right now, which means it really is a SPOSet and too much other stuff. It's sad to bake this state machine further into the SPOSet infrastructure and reinforce an API member that shouldn't be in SPOSet.

A RotatedSPOSet doesn't have a is-a relationship to an SPOSet. It requires the definition of an SPOSet to be changed to have an API member `storeParamsBeforeRotation` that is part of being a "SPOSet" but actually is a noop in all cases (7+BsplineSet derived classes) except for LCAOOrbitalSet.

Even if we are going to have such an awkward object shouldn't it have a const reference to the original SPOSet and the rotation that has been applied.  And then for optimization you could add habe it be derived from the derived type of the original SPOSet and operate on its own coefficients instead of making an SPOSet capable of holding two sets of coefficient data occassionaly for some derived types.

Reply:
BsplineSet needs to implement storeParamsBeforeRotation if optimizing orbital rotations is going to work on splines.

I tried implementing an SPOSetRotatable class that derives from SPOSet and adds storeParamsBeforeRotation and applyRotation.  The idea is that only SPO sets that implement rotation would derive from SPOSetRotatable (so LCAOrbitalSet and BsplineSet).  And RotatedSPOs would take an SPOSetRotatable object as input (instead of the current SPOSet).   I couldn't get it to compile (issues related to dynamic casting SPOSet to SPOSetRotatable in setup).

At the most concrete level, some piece of code needs to multiply the rotation matrix by the coefficient matrix.  It makes sense to put this at the LCAOrbitalSet level. Or some other piece of code will need to know about LCAOrbitalSet internals to do that multiplication.

Applying a delta rotation introduces more effective state into an object - the number of times the delta rotation has been applied (and the particular values of each rotation)

Reply:
RotatedSPO should carry some states "delta rotation" and "total rotation"(not there yet) but it should not know what is the actual underlying SPOSet. However, it also needs some help from the SPOSet. To make "total rotation" meaningful, SPOSet is responsible for the original coefficients.

I have doubts in the necessity of SPOSetRotatable intermediate base class.